crates/oxc_semantic/src/builder.rs

```diff
@@ -100,18 +95,103 @@ pub struct SemanticBuilderReturn<'a> {
     pub errors: Vec<OxcDiagnostic>,
 }
 
+// Stack used to accumulate unresolved refs while traversing scopes.
+// Indexed by scope depth. We recycle `UnresolvedReferences` instances during traversal
+// to reduce allocations, so the stack grows to maximum scope depth, but never shrinks.
+// See: <https://github.com/oxc-project/oxc/issues/4169>
+// This stack abstraction uses the invariant that stack only grows to avoid bounds checks.
+struct UnresolvedReferencesStack {
```

Comment:
Move to a new file? Seems out of place.

Reply:
Done. NB this is only used in `SemanticBuilder`, not `Semantic`, so I don't think it should be in `scope.rs`.